### PR TITLE
DRILL-7525: Convert SequenceFiles to EVF

### PR DESCRIPTION
**Pull request by luocooong**
# [DRILL-7525](https://issues.apache.org/jira/browse/DRILL-7525): Convert SequenceFiles to EVF

## Description

 Converts the SequenceFiles format plugin to use EVF.

## Documentation
 There are no changes visible to the user.

## Testing
 Ran unit tests associated with the SequenceFiles plugin. Added tests method for other.

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileBatchReader.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+  @Override
+  public boolean next() {
+    int recordCount = 0;
+    if (watch == null) {
+      watch = Stopwatch.createStarted();
+    }
+    try {
+      while (!loader.isFull()) {
+        if (reader.next(key, value)) {
```
Comment (cgivre) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileBatchReader.java:138]: I strongly recommend adding the limit pushdown code.   All you need to do is add the code I referred to earlier and then in this loop add the following:
```java
if (loader.limitReached(maxRecords) ) {
  return false;
}

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileBatchReader.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+          ++ recordCount;
+        } else {
+          logger.debug("Read {} records in {} ms", recordCount, watch.elapsed(TimeUnit.MILLISECONDS));
+          return false;
+        }
+      }
+    } catch (IOException e) {
+      close();
+      throw UserException
+              .dataReadError(e)
```
Comment (cgivre) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileBatchReader.java:152]: Please make sure you add a message as to what went wrong here.    You'll want to make sure every error message has:
* a clear message in the `message()` call and
* a call to the `addContext(errorContext)`

Here and elsewhere.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileBatchReader.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+          valueWriter.setBytes(value.getBytes(), value.getLength());
+          loader.save();
+          ++ recordCount;
+        } else {
+          logger.debug("Read {} records in {} ms", recordCount, watch.elapsed(TimeUnit.MILLISECONDS));
+          return false;
+        }
+      }
+    } catch (IOException e) {
+      close();
```
Comment (cgivre) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileBatchReader.java:150]: The `close()` call is not needed here.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileBatchReader.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+              .addContext("File Path", split.getPath())
+              .build(logger);
+    }
+    return true;
+  }
+
+  @Override
+  public void close() {
+    try {
+      if (reader != null) {
```
Comment (cgivre) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileBatchReader.java:162]: Does the reader implement `AutoCloseable`?  If so you might want to use that instead.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileBatchReader.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        reader.close();
+        reader = null;
+      }
+    } catch (IOException e) {
+      logger.warn("Exception closing reader: {}", e);
+    }
+  }
+
+  @Override
+  public String toString() {
```
Comment (cgivre) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileBatchReader.java:172]: The `toString()` function is not really needed for the BatchReader classes.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileBatchReader.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    try {
+      if (reader != null) {
+        position = reader.getPos();
+      }
+    } catch (IOException e) {
+      logger.trace("Unable to obtain reader position.", e);
+    }
+    return String.format("SequenceFileBatchReader[File=%s, Position=%s]", split.getPath(), position);
+  }
+
```
Comment (cgivre) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileBatchReader.java:183]: NIT:  Remove extra line.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileFormatPlugin.java [hunk 7]
```diff
@@ -17,92 +17,77 @@
...
+    config.readable = true;
+    config.writable = false;
+    config.blockSplittable = true;
+    config.compressible = true;
+    config.extensions = pluginConfig.getExtensions();
+    config.fsConf = fsConf;
+    config.readerOperatorType = CoreOperatorType.SEQUENCE_SUB_SCAN_VALUE;
+    config.useEnhancedScan = true;
+    config.supportsProjectPushdown = true;
+    config.defaultName = SequenceFileFormatConfig.NAME;
```
Comment (cgivre) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/sequencefile/SequenceFileFormatPlugin.java:56]: If you want to support the limit pushdown, you should add
```
config.supportsLimitPushdown = true;
```
to this method.

File: exec/java-exec/src/test/java/org/apache/drill/exec/store/sequencefile/TestSequenceFileReader.java [hunk 8]
```diff
@@ -17,32 +17,59 @@
...
+  }
+
+  private static String byteWritableString(String input) throws Exception {
+    final ByteArrayOutputStream bout = new ByteArrayOutputStream();
+    DataOutputStream out = new DataOutputStream(bout);
+    final BytesWritable writable = new BytesWritable(input.getBytes("UTF-8"));
+    writable.write(out);
+    return new String(bout.toByteArray());
   }
+
```
Comment (cgivre) [inline exec/java-exec/src/test/java/org/apache/drill/exec/store/sequencefile/TestSequenceFileReader.java:74]: I'd recommend some additional work on the unit tests.  First, please remove any code in the unit tests that produces output to STDOUT.  (IE any print statements) 

Secondly, take a look at the test below:  I'd recommend refactoring all the unit tests for this in the pattern below. This pattern will test that the schema is being generated correctly, and that the outputs are properly mapped using EVF.

I'd also add unit tests for:
* Star Query
* Explicit Field Query
* Limit Pushdown (Plan only)
* SerDe 
* Compressed File (If applicable)

You can pretty much cut/paste from the file below for all those tests. 


https://github.com/apache/drill/blob/f79587ed14767463129f8905d2e36e55563655f2/contrib/format-spss/src/test/java/org/apache/drill/exec/store/spss/TestSpssReader.java#L110-L131

**Review by cgivre — CHANGES_REQUESTED** [review]
@luocooong 
Thanks for the PR!  Nice work!  I left some review comments for you to consider.  One question I had is that I saw that sequence files do have metadata.  Would it be of any value to a user to include this metadata as an implicit field in Drill?  If so, I can show you how to do that.

**Comment by luocooong** [comment]
@cgivre All done. Thanks.

**Review by cgivre — CHANGES_REQUESTED** [review]
@luocooong 
Thanks for the responses.  Minor thing... Can you please make sure all exception are in 

```
throw UserException
               .dataReadError(e)
               .message(<plain message> + e.getMessage() )
               .addContext(errorContext)
               .build(logger);
```
One of the challenges of Drill are difficult error messages, so we're trying to make sure that if something goes wrong, we give the user enough information to actually fix the issue rather than dumping the stack trace.

The exceptions which already have a message are fine, but just please add a plain text explanation of what went wrong.  Once that's done for all exceptions, please squash commits, and we're good to go.

**Review by cgivre — APPROVED** [review]
LGTM +1
Please squash all commits and thanks for the contribution!!

**Pull request merged**